### PR TITLE
Add OpenStack Support

### PR DESCRIPTION
**Pull request by jimkoen**
This adds support for OpenStack as a DNS provider

### CODE REVIEW

File: roles/acme_certificate/tasks/dns-openstack-cleanup.yml [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+# GNU General Public License v3.0+ (see LICENSES/GPL-3.0-or-later.txt or https://www.gnu.org/licenses/gpl-3.0.txt)
+# SPDX-License-Identifier: GPL-3.0-or-later
+# SPDX-FileCopyrightText: 2022, Lucca Jiménez Könings (https://github.com/jimkoen)
+
+# Clean up DNS challenges for DNS provider Openstack
+- name: Cleaning up challenge DNS entries for {{ ', '.join(acme_certificate_domains + acme_certificate_ips) }} via OpenStack
+  openstack.cloud.recordset:
+    state: absent
+    # Openstack expects a zone name with a trailing dot
+    zone: "{{ acme_certificate_openstack_zone | default(item.key | community.dns.get_registrable_domain) }}."
```
Comment (felixfontein) [inline roles/acme_certificate/tasks/dns-openstack-cleanup.yml:11]: What is `acme_certificate_openstack_zone`? I don't see it documented anywhere.

Also when using `item.key`, you need to use `felixfontein.acme._substitute_dns(acme_certificate_dns_substitution)` as well (see the other DNS providers in the `main` branch).

File: roles/acme_certificate/tasks/dns-openstack-cleanup.yml [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+    records: "{{ item.value }}"
+    recordset_type: txt
+    ttl: 60
+    name: "{{ item.key }}."
```
Comment (felixfontein) [inline roles/acme_certificate/tasks/dns-openstack-cleanup.yml:15]: Please check the other DNS providers in the `main` branch, you need to use the `felixfontein.acme._substitute_dns(acme_certificate_dns_substitution)` filter here.

File: roles/acme_certificate/tasks/dns-openstack-create.yml [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+# Create DNS challenges for DNS provider Openstack
+- name: Creating challenge DNS entries for {{ ', '.join(acme_certificate_domains + acme_certificate_ips) }} via OpenStack
+  openstack.cloud.recordset:
+    state: present
+    # Openstack expects a zone name with a trailing dot
+    zone: "{{ acme_certificate_openstack_zone | default(item.key | community.dns.get_registrable_domain) }}."
+    records: "{{ item.value }}"
+    recordset_type: txt
+    ttl: 60
+    name: "{{ item.key }}."
```
Comment (felixfontein) [inline roles/acme_certificate/tasks/dns-openstack-create.yml:15]: Same here.

File: roles/acme_certificate/tasks/dns-openstack-sanity.yml [hunk 4]
```diff
@@ -0,0 +1,31 @@
+---
+# GNU General Public License v3.0+ (see LICENSES/GPL-3.0-or-later.txt or https://www.gnu.org/licenses/gpl-3.0.txt)
+# SPDX-License-Identifier: GPL-3.0-or-later
+# SPDX-FileCopyrightText: 2022, Lucca Jiménez Könings (https://github.com/jimkoen)
+
+- name: DNS Sanity Checks (Openstack)
+  block:
+    - openstack.cloud.config:
```
Comment (felixfontein) [inline roles/acme_certificate/tasks/dns-openstack-sanity.yml:8]: This should only be run *after* testing whether that module exists.

File: roles/acme_certificate/tasks/dns-openstack-sanity.yml [hunk 5]
```diff
@@ -0,0 +1,31 @@
+---
+# GNU General Public License v3.0+ (see LICENSES/GPL-3.0-or-later.txt or https://www.gnu.org/licenses/gpl-3.0.txt)
+# SPDX-License-Identifier: GPL-3.0-or-later
+# SPDX-FileCopyrightText: 2022, Lucca Jiménez Könings (https://github.com/jimkoen)
+
+- name: DNS Sanity Checks (Openstack)
+  block:
```
Comment (felixfontein) [inline roles/acme_certificate/tasks/dns-openstack-sanity.yml:7]: Is the block needed for anything else than the `openstack.cloud.config` module?

File: roles/acme_certificate/tasks/dns-openstack-sanity.yml [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+          - "'openstack.cloud.recordset' is community.general.a_module"
+        msg: "The openstack.cloud collection with the openstack.cloud.recordset module must be installed for Openstack DNS"
+      run_once: true
+
+    - name: DNS sanity checks (2/2)
+      ansible.builtin.assert:
+        that:
+          - (openstack.clouds | length) >= 1
+          - (openstack.clouds | rejectattr('auth_type', 'equalto', 'password') | list | length) == 0
+        msg: "For Openstack DNS, auth plugins other than the default 'password' plugin are not supported at this time."
```
Comment (felixfontein) [inline roles/acme_certificate/tasks/dns-openstack-sanity.yml:24]: `openstack.clouds` is not mentioned in the documentation.

File: roles/acme_certificate/tasks/dns-openstack-sanity.yml [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+      run_once: true
+  rescue:
+    - ansible.builtin.debug:
+        msg: "The provided OpenStack auth configuration (either via clouds.yaml or environment variables) is not valid."
+
+    
+
```
Comment (felixfontein) [inline roles/acme_certificate/tasks/dns-openstack-sanity.yml:31]: There are too many blank lines here.

**Review by felixfontein — COMMENTED** [review]
Thanks for your contribution!

- Please add a [changelog fragment](https://docs.ansible.com/ansible/devel/community/collection_development_process.html#creating-a-changelog-fragment).
- This should also be mentioned in the documentation, in particular in docs/docsite/rst/general_role_parameters.rst.

**Comment by jimkoen** [comment]
I'm still comitted on fixing this, just busy atm.

**Comment by jimkoen** [comment]
I'm writing exams right now though. I'll likely have this done before march.

**Comment by felixfontein** [comment]
@jimkoen take your time, no need to hurry :)

**Comment by jimkoen** [comment]
Regarding `acme_certificate_openstack_zone`:

This parameter poses a bit of an issue in combination with the DNS CNAME substitution and the use of `community.dns.get_registrable_domain`.

We run Openstack in a multi-tenant evironment. As such, each user gets assigned a DNS Zone that is a Subdomain of our parent DNS Zone. To give an example, our cloud runs under `myOpenstackCloud.com` and each user is assigned a private cloud prefix for their project under `*.users.myOpenstackCloud.com`. For example, the prefix for my project would be
`jimkoen.users.myOpenstackCloud.com`.

I create DNS entries via the Openstack Ansible Modules, [which require a zone name parameter.](https://docs.ansible.com/ansible/latest/collections/openstack/cloud/dns_zone_module.html#return-zone/name) This parameter is the general zone for the user-prefix, as well as the registerable domain part. In my example this would be `jimkoen.users.myOpenstackCloud.com.`,
with a dotted suffix. Currently, I use the zone name and whatever record I want to add and concatenate them together. So for `my.host.jimkoen.users.myOpenstackCloud.com` I'd be `"{{ my.host }}. {{ jimkoen.users.myOpenstackCloud.com }}."` as a Jinja2 template for the zone name.

This somewhat clashes with the general design of this module, as well as the community modules, because they all expect the DNS Zone to be a top level domain, instead of a subdomain. I looked into how to use your `felixfontein.acme._substitute_dns` filter, but I think this is intended only for CNAMES? I might be able to hack something together with that though.

I will of course add documentation for the `acme_certificate_openstack_zone` parameter, but I'm not sure how DNS substitution would work in our environment, with how it's intended in the module.

**Comment by jimkoen** [comment]
@felixfontein 
Unfortunately my previous employer and myself are going separate ways.
Due to my departure I no longer have access to a working OpenStack Cloud to test development with.
Which is why I'm closing the PR now. Sorry :/

**Comment by felixfontein** [comment]
@jimkoen sorry for not replying to your earlier message, I must have missed it...

> Unfortunately my previous employer and myself are going separate ways.
> Due to my departure I no longer have access to a working OpenStack Cloud to test development with.
> Which is why I'm closing the PR now. Sorry :/

That's really unfortunate, but that's life :) Thank you very much for all your work on this!

**Pull request closed without merging**